sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -39,7 +41,15 @@ object SimplifyExtractValueOps extends Rule[LogicalPlan] {
       // Remove redundant field extraction.
       case GetStructField(createNamedStruct: CreateNamedStruct, ordinal, _) =>
         createNamedStruct.valExprs(ordinal)
-
+      case GetStructField(w@WithFields(struct, names, valExprs), ordinal, maybeGetFieldName) =>
+        val getFieldName = maybeGetFieldName.getOrElse(w.dataType(ordinal).name)
```

Comment:
I think it's better to use `w.dataType(ordinal).name` only, then we can write `names.zip(valExprs).filter(_._1 == fieldName)` to find the match.

Reply:
you're right, done.